### ISSUE TITLE
Cypress Doesn't Fail Test on (Unhandled) Promise Rejection

### ISSUE THREAD

**Issue by andezzat** [cypress-io/cypress#2487]
### Current behavior:
![screen shot 2018-09-18 at 3 50 19 pm](https://user-images.githubusercontent.com/1734293/45666982-a5540d80-bb5a-11e8-8db5-2a5bdead0bf6.png)

Cypress passes test, although the HTML page contains a script in the head, here's the full markup:
```
<html lang="en">
  <head>
    <title>Hello World</title>
    <script>
      (async function () {
        const x = 1;
        throw new Error('error!');
      })()
    </script>
  </head>
  <body>
    <div>
      <h1>This should fail...</h1>
    </div>
  </body>
</html>
```

### Desired behavior:

Cypress should fail the test if there's an unhandled promise rejection?

### Steps to reproduce:

Setup Cypress in a new repo, serve a static Index.html file with a script in the head that throws an error from inside an async function or a promise.
Invoke said async/promise in your code & then run a Cypress test on that page. It will pass!

### Versions

Cypress 3.1.0
Electron 59
macOS 10.13.6

**Comment by jennifer-shehane** [comment]
Was able to reproduce. Added this test to `cypress-example-kitchensink` app:

```js
it('should fail', () => {
  cy.visit('app/index.html')
})
```

Add this to `index.html`:

```html
  <script>
    (async function () {
      const x = 1;
      throw new Error('error!');
    })()
  </script>
```

**Comment by marcelo-rebello** [comment]
Is there any way to fail a test based on console errors?

**Comment by andezzat** [comment]
@marcelo-rebello I was struggling with this for a while and found my own fix..
The real question is: why can't you just setup your code to throw an error onto the window?
If the answer (same as me) is because you already have a function to catch uncaught errors on `window.onerror` then you need to somehow disable that function from being assigned in your code while Cypress is running...
That was my fix.

Otherwise, if you STILL want to fail on console.log and DON'T have a function assigned to `window.onerror`, then by all means use `cy.stub` to stub your `console.error` and force it to fail by throwing said error msg..
This should technically work:
```
cy
  .window()
  .then(win => cy.stub(win , 'console.error', msg =>
    setTimeout(() => { throw new Error(msg); })));
```

`setTimeout` guarantees that if throwing inside an async function, your error will be caught by the window (uncaught error). Experiment and see what works for you. Good luck!

**Comment by marcelo-rebello** [comment]
This stub didn't work for me @andezzat. Cypress logs the error, but the test doesn't fail...

The best "solution" until now is to invoke a `cy.task('error')`  that just `return undefined`.
This makes Cypress raise another error because I am invoking a Command inside a Promise,
but then I can ignore this error and the test will fail (pretty ugly solution, I know)

I suppose this happens because of the exotic sync/async/Promisse/Serialize implementation of Cypress, but if Errors are so hard to be detected, a `cy.fail()` command would be a great add on...

**Comment by andezzat** [comment]
I would not recommend this approach as you're depending on a completely unrelated behaviour of Cypress to fail your tests... this could easily change in the future and as such is not a reliable way.

I'll experiment a bit and see if there's another way..
Why can't you just have your app throw the error instead of logging it btw?
@marcelo-rebello

**Comment by marcelo-rebello** [comment]
I completely agree that this is not a good solution, nor it is safe to work in future, but I didn't find any other way :(
If a throw an error on Vue error handler function, it is logged, but Cypress test didn't fail...
and I really want that some Vue warns/errors make the tests fail.
As I stubbed `console.error`, I only manage to test failing using Task error :(

**Comment by andezzat** [comment]
@marcelo-rebello I'm not entirely familiar with Vue's error handler (despite actually using Vue here at work)..
Can you try throwing your errors to the window like so to see if it helps?
`setTimeout(() => { throw new Error('woopsie'); });`

**Comment by marcelo-rebello** [comment]
just don't work, I had already tried...
I put that code at the end of Vue error handler function and the test succeeds :(
But I can see 2 errors on the console related to this exception:
```
VM2589 utils.js:36 Uncaught Error: woopsie at eval (VM1028 utils.js:36)
```
and
```
Uncaught Error: Script error.

Cypress detected that an uncaught error was thrown from a cross origin script.

We cannot provide you the stack trace, line number, or file where this error occurred.

Check your Developer Tools Console for the actual error - it should be printed there.

It's possible to enable debugging these scripts by adding the 'crossorigin' attribute and setting a CORS header.

https://on.cypress.io/cross-origin-script-error

This error originated from your application code, not from Cypress.

When Cypress detects uncaught errors originating from your application it will automatically fail the current test.

This behavior is configurable, and you can choose to turn this off by listening to the 'uncaught:exception' event.

https://on.cypress.io/uncaught-exception-from-application
    at createErrFromMsg (cypress_runner.js:56902)
    at Object.createUncaughtException (cypress_runner.js:56909)
    at Object.onUncaughtException (cypress_runner.js:59793)
    at onError (cypress_runner.js:59292)
```

**Comment by andezzat** [comment]
@marcelo-rebello I see... interesting. You might want to create a separate issue for this I guess?

**Comment by marcelo-rebello** [comment]
@andezzat looks like the same error of this issue (and it still open), some Exceptions are being detected by Cypress, but not failing the test. (and there is no good way to force Cypress to fail)

**Comment by andezzat** [comment]
Ah I see what you mean now. You are indeed throwing the error, just Cypress not catching on...
To help this Issue move forward faster, see if you can reproduce YOUR specific problem with a very minimal setup...
@marcelo-rebello

**Comment by andezzat** [comment]
Updated issue to add more cases where Cypress does not fail on unhandled exceptions.
@jennifer-shehane fyi

**Comment by jennifer-shehane** [comment]
Hey @andezzat, I have an update on this issue. 

Currently, Chrome is the only browser that throws `unhandledrejection` events instead of `error` events. Cypress only listens to `error` events to fail the test. However, in 4.0, we've extended Cypress to also listen to `unhandledrejection` events, so be on the lookout for [4.0 release](https://github.com/cypress-io/cypress/issues/2840). https://github.com/cypress-io/cypress/issues/2884

**Comment by andezzat** [comment]
Thanks for the update @jennifer-shehane.. that makes sense. Hopefully 4.0 is not too long around the corner :)

**Comment by ThomasGHenry** [comment]
Running 4.0.1 and unhandled rejections are still passing.

**Comment by garcianavalon** [comment]
Same here, errors thrown in my Vue app get shown logged the console but Cypress won't fail the test. Is there any workaround for this behaviour?

**Comment by ThomasGHenry** [comment]
Hopefully, there's a better way, but I ended up asserting on visual differences (react error page vs working page).

**Comment by bahmutov** [comment]
Do you have a reproduction? I’m thinking the error is thrown after the test finishes

**Comment by garcianavalon** [comment]
I can build a minimal Vue app that throws and error and a cypress test, I'll get back to you

**Comment by bahmutov** [comment]
I have tried it and confirmed the test finishes too quickly - the error is thrown _after_ the test. We have described this in the blog post https://www.cypress.io/blog/2020/01/16/when-can-the-test-stop/

If you suspect the page throws an error, the simplest way is to add a wait at the end

```js
/// <reference types="cypress" />
it('should fail', () => {
  cy.visit('index.html')
      .wait(100)
})
```

which fails the test correctly.

<img width="1440" alt="Screen Shot 2020-02-20 at 2 01 54 PM" src="https://user-images.githubusercontent.com/2212006/74968942-b9eca300-53e9-11ea-9b3c-709f706f6d9d.png">

@jennifer-shehane is there anything else we can do?

**Comment by WayneHiller** [comment]
@garcianavalon Yes as far as Vue is concerned this is the same issue I noted in #5385. It is not a bug in Cypress but the design of Vue. Vue will handle exceptions itself, even when thrown inside a Vue error handler so cypress never sees them.

**Comment by shivasrini** [comment]
> I have tried it and confirmed the test finishes too quickly - the error is thrown _after_ the test. We have described this in the blog post https://www.cypress.io/blog/2020/01/16/when-can-the-test-stop/
> 
> If you suspect the page throws an error, the simplest way is to add a wait at the end
> 
> ```js
> /// <reference types="cypress" />
> it('should fail', () => {
>   cy.visit('index.html').wait(100)
> })
> ```
> 
> which fails the test correctly.
> 
> <img alt="Screen Shot 2020-02-20 at 2 01 54 PM" width="1440" src="https://user-images.githubusercontent.com/2212006/74968942-b9eca300-53e9-11ea-9b3c-709f706f6d9d.png">
> 
> @jennifer-shehane is there anything else we can do?

@bahmutov I am also facing the same issue where the test is possibly finishing before the error is thrown. Any viable workarounds for this yet?

**Comment by bahmutov** [comment]
@shivasrini please fix the application code or add a wait to the test to ensure that any potential error thrown fails the test. Cypress has no idea if and when your application code might throw an error, so it won't by default wait for it.

**Comment by JohnnyFun** [comment]
I modified the kitchen sink app to do what @jennifer-shehane did in her test above, but in the context of a test that takes a while to run. The error is thrown and logged to the console long before the test completes. Should this be caught by cypress's `unhandledrejection` handler and fail the test, @bahmutov? Or am I misunderstanding something?

![unhandled](https://user-images.githubusercontent.com/8106379/78944165-8558a780-7a7a-11ea-98d4-7d10e8cd8077.gif)

Looks like kitchen sink is running cypress@4.2.0, fwiw. Let me know if you need me to share anything else.

**Comment by bahmutov** [comment]
Give us the code to do this to look

**Comment by JohnnyFun** [comment]
Sure thing, I cloned the repo and made the change:

Only added this script tag to the actions spec:
![image](https://user-images.githubusercontent.com/8106379/78947167-e46dea80-7a81-11ea-932f-14539ac63ac9.png)

So:
- `git clone https://github.com/JohnnyFun/cypress-example-kitchensink`
- `cd cypress-example-kitchensink`
- `npm i`
- `npm run start:ci:windows`
  - or `npm run start:ci` or `npm start` if you're not windows

Then in another prompt run: `npm run cy:open`
And run `actions.spec.js` with dev tools open.
Observe that the exceptions is logged before the test completes.

Thanks Gleb! No rush, btw. Not a show stopper or anything.

**Comment by goldylucks** [comment]
any news on this?

**Comment by antur84** [comment]
I stumbled in here while having very similar issues when I wanted to be failing tests if console.error is logged, and just wanted to share the workaround that solves our problems related to this.

```js
/**
 * Fail the test run if there's an error in the console.
 */
Cypress.on('window:before:load', win => {
    win.console.error = msg => {
        // Schedule an unhandled exception in the stack instead of throwing it directly.
        // This is a workaround for unstable cypress behavior when
        // there's an errors in the console but the tests just keeps on running
        setTimeout(() => {
            throw new Error(msg);
        }, 0);
    };
});
```

Without the timeout, errors logged late in each spec would not "be seen" by cypress, so no test failure happens. Errors happening early on would be seen and tests would fail

**Comment by Kondziowsky** [comment]
any solution?

i need to ignore request for .ttf file 404 ERR
